app/models/metasploit/credential/krb_enc_key.rb

```diff
@@ -0,0 +1,184 @@
+  #
+
+  validate :data_format
+
+  #
+  # Class methods
+  #
+
+  # @param [Integer] enctype The enctype
+  # @param [String] key The key bytes
+  # @param [String,nil] salt The salt
+  # @return [String]
+  # @raise [ArgumentError] if an option is invalid
+  def self.build_data(enctype:, key:, salt: nil)
+    raise ArgumentError('enctype must be numeric') unless enctype.is_a?(Numeric)
+    raise ArgumentError('key must be set') if key.nil?
+
+    "msf_krbenckey:#{enctype}:#{as_hex(key)}:#{as_hex(salt)}"
+  end
+
+  #
+  # Instance Methods
+  #
+
+  # The enctype as defined by https://www.iana.org/assignments/kerberos-parameters/kerberos-parameters.xhtml
+  #
+  # @return [Integer]
+  def enctype
+    parsed_data[:enctype]
+  end
+
+  # The key
+  #
+  # @return [String]
+  def key
+    parsed_data[:key]
+  end
+
+  # The salt used as part of creating the key. This is normally derived from the Kerberos principal name/Realm.
+  # For windows the following convention is used to create the salt:
+  # https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-kile/7a7b081d-c0c6-46f4-acbf-a439664270b8
+  #
+  # This value can be nil if the salt is not known
+  # @return [String,nil] The key salt if available
+  def salt
+    parsed_data[:salt]
+  end
+
+  # A string suitable for displaying to the user
+  #
+  # @return [String]
+  def to_s
+    "#{ENCTYPE_NAMES[enctype]}:#{self.class.as_hex(key)}#{salt ? ":#{self.class.as_hex(salt)}" : ''}"
+  end
+
+  private
+
+  # Converts a buffer containing bytes to a String containing the hex representation of the bytes
+  #
+  # @param hash [String,nil] a buffer of bytes
+  # @return [String] a string where every 2 hexadecimal characters represents a byte in the original hash buffer
+  def self.as_hex(value)
+    value.to_s.unpack1('H*')
+  end
+
+  # Converts a buffer containing bytes to a String containing the hex representation of the bytes
+  #
+  # @param hash [String,nil] a buffer of bytes
+  # @return [String] a string where every 2 hexadecimal characters represents a byte in the original hash buffer
+  def self.as_bytes(value)
+    [value.to_s].pack('H*')
+  end
+
+  # @return [Hash] The parsed data with enctype, key, salt keys
+  def parsed_data
+    match = data.match(DATA_REGEXP)
+    return {} unless match
+
+    {
+      enctype: match[:enctype].to_i,
+      key: self.class.as_bytes(match[:key]),
+      salt: match[:salt].empty? ? nil : self.class.as_bytes(match[:salt])
+    }
+  end
+
+  # Normalizes {#data} by making it all lowercase so that the unique validation and index on
+  # ({Metasploit::Credential::Private#type}, {#data}) catches collision in a case-insensitive manner without the need
+  # to use case-insensitive comparisons.
+  def normalize_data
+    if data
+      self.data = data.downcase
+    end
+  end
+
+  # Validates that {#data} is in the expected data format
+  def data_format
+    unless DATA_REGEXP.match(data)
+      errors.add(:data, :format)
+    end
+  end
```

Comment:
Is it possible to make this more descriptive? Maybe expand this to determine what is `incorrect` about the format to provide better debugging data than `this did not match the expected regex`.

Reply:
At runtime this will get mapped to the i18n translations here https://github.com/rapid7/metasploit-credential/pull/168/files#diff-44438ce218f5287c58d0017f965d888715635d94280669896f75841fbd7b4cd7

So the real user error will be: `is not in the KrbEncKey data format of 'msf_krbenckey:<ENCTYPE>:<KEY>:<SALT>', where the key and salt are in hexadecimal characters`

In the future we might add extra checks for key lengths though, as rc4/aes128/aes256 are all 32/64 bytes